File: leetcode/Python/0140-h-word-break-ii.py
"""
Recursion with memoization
"""
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        return self.breakW(s, {}, set(wordDict))
    
    def breakW(self, s, mem, words):
        if s in mem:
            return mem[s]
        res = []
        if s in words:
            res.append(s)
        for i in range(1, len(s)):
            left = s[:i]
            right = s[i:]
            if right not in words:
                continue
            res += [w + " " + right for w in self.breakW(left, mem, words)]
        mem[s] = res
        return mem[s]

File: leetcode/C++/0140-h-word-break-ii.cpp
// Recursion with memorization, TLE
class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        set<string> words(wordDict.begin(), wordDict.end());
        unordered_map<string, vector<string>> mem;
        return breakW(s, mem, words);
    }

    vector<string> breakW(string s, unordered_map<string, vector<string>> mem, set<string> words) {
        if (mem.find(s) != mem.end())
            return mem[s];
        
        vector<string> res;
        if (words.find(s) != words.end())
            res.push_back(s);
        for (int i = 1; i < s.length(); i++) {
            string left = s.substr(0, i);
            string right = s.substr(i, s.length() - i);
            if (words.find(right) == words.end())
                continue;
            vector<string> rem = breakW(left, mem, words);
            for (string w : rem)
                res.push_back(w + " " + right);
        }
        mem[s] = res;
        return mem[s];
    }
};